### PR TITLE
Vardt new

### PR DESCRIPTION
**Pull request by babrodtk**
Can you take a look at this @setmar ?

### CODE REVIEW

**Comment by babrodtk** [comment]
It has not been tested with your additions by me...

File: python/SWESimulators/Common.py [hunk 1]
```diff
@@ -55,18 +54,24 @@ def __init__(self, cl_ctx, nx, ny, halo_x, halo_y, data, \
...
+            assert(host_data.shape[0] == self.ny_halo), str(host_data.shape[0]) + " vs " + str(self.ny_halo)
+            assert(self.bytes_per_float == 4)
+            assert(data.shape == (self.ny_halo, self.nx_halo))
+        else:
+            self.bytes_per_float = 4
+            self.data = pyopencl.Buffer(cl_ctx, mf.READ_WRITE, self.nx_halo*self.ny_halo*self.bytes_per_float)
         
-        self.bytes_per_float = host_data.itemsize
-        assert(self.bytes_per_float == 4)
+        self.holds_data = True
```
Comment (babrodtk) [inline python/SWESimulators/Common.py]: Why is this here? What does it signify? Looks scary to me... Shouldn't these always own their own data?

File: python/SWESimulators/Simulator.py [hunk 2]
```diff
@@ -65,7 +65,10 @@ def __init__(self, \
         self.ghost_cells_y = np.int32(ghost_cells_y)
         self.dx = np.float32(dx)
         self.dy = np.float32(dy)
-        self.dt = np.float32(dt)
+        if dt is not None:
+            self.dt = np.float32(dt)
+        else:
+            self.dt = dt
```
Comment (babrodtk) [inline python/SWESimulators/Simulator.py]: This is always self.dt = dt, or no?

Comment (babrodtk) [reply python/SWESimulators/Simulator.py]: Code should be made a bit more readable here.

File: python/VariableDt.ipynb [hunk 3]
```diff
@@ -0,0 +1,4277 @@
+{
```
Comment (babrodtk) [inline python/VariableDt.ipynb]: 4277 new lines? Has this file been cleaned before checkin?

File: python/SWESimulators/Common.py (hunk 1 above)
Comment (setmar) [reply python/SWESimulators/Common.py]: I think this is to avoid free'ing uninitialized memory on GPU. This was not added in this PR.

File: python/SWESimulators/KP07.py [hunk 4]
```diff
@@ -215,117 +257,130 @@ def cleanUp(self):
...
+
+        n = 0
+        t_end = np.float32(t_end);
+        while (t_end > 0.0):
+            if (self.use_rk2):
+                # Substep one
+                self.fluxKernel(update_dt, self.cl_data.h0, self.cl_data.hu0, self.cl_data.hv0)
+                if self.dt is not None:
+                    local_dt = np.float32(min(t_end, local_dt))
+                    upload_dt = np.zeros((128), dtype=np.float32)
```
Comment (babrodtk) [inline python/SWESimulators/KP07.py]: Why hardcoded 128 here? Should this be dt_block_size or similar?

File: python/VariableDt.ipynb (hunk 3 above)
Comment (setmar) [reply python/VariableDt.ipynb]: Probably not? Are you using ipynb_drop_output script?

File: python/SWESimulators/KP07.py [hunk 5]
```diff
@@ -215,117 +257,130 @@ def cleanUp(self):
...
-                self.bc_kernel.boundaryCondition(self.cl_queue, \
-                        self.cl_data.h0, self.cl_data.hu0, self.cl_data.hv0)
                 
-            self.t += local_dt
+                        
+            else:
+                self.fluxKernel(update_dt, self.cl_data.h0, self.cl_data.hu0, self.cl_data.hv0)
+                if self.dt is not None:
+                    local_dt = np.float32(min(t_end, local_dt))
+                    upload_dt = np.zeros((128), dtype=np.float32)
```
Comment (babrodtk) [inline python/SWESimulators/KP07.py]: Again, why 128 hardcoded here?

File: python/SWESimulators/KP07.py [hunk 6]
```diff
@@ -215,117 +257,130 @@ def cleanUp(self):
...
+                self.boundaryConditionsKernel(self.cl_data.h1, self.cl_data.hu1, self.cl_data.hv1)
+                      
+                # Swap h0 and h1  
+                self.cl_data.swap()
+            
+            if self.dt is not None:
+                t_end = t_end - local_dt;
+                self.t += np.float32(local_dt);
+            else:
+                local_dt = np.zeros((128), dtype=np.float32)
```
Comment (babrodtk) [inline python/SWESimulators/KP07.py]: Again, why 128 here?

File: python/SWESimulators/KP07.py (hunk 4 above)
Comment (babrodtk) [reply python/SWESimulators/KP07.py]: I think this should be just 1, and was set to 128 for testing.

Comment (setmar) [reply python/SWESimulators/KP07.py]: Probably. Uncritical copy-pasting from my part.

File: python/SWESimulators/Common.py (hunk 1 above)
Comment (havahol) [reply python/SWESimulators/Common.py]: That's right! I added self.holds_data to avoid freeing twice (which would give an exception)

**Comment by babrodtk** [comment]
This PR still runs more kernels than before, and will therefore impose a significant performance penalty. It also uses far more storage than originally. 

I am still not certain that it is a very good idea to merge this into master. Perhaps we'll just keep it as a separate commit here (though, then all of the nasty whitespace-changes should be removed - to keep the diff clean and readable for future cherry picking...)

File: python/SWESimulators/Simulator.py (hunk 2 above)
Comment (babrodtk) [reply python/SWESimulators/Simulator.py]: (though this is a minor issue ;) )

File: python/VariableDt.ipynb (hunk 3 above)
Comment (babrodtk) [reply python/VariableDt.ipynb]: I didn't touch the file, so I don't know. If you are using that script, it probably is fine.

File: python/SWESimulators/KP07.py (hunk 4 above)
Comment (babrodtk) [reply python/SWESimulators/KP07.py]: And uncritical commit from me. No biggie, but should be fixed in the future. Will crash for small domains, I think (if the device size buffer is less than 128 long)

**Comment by babrodtk** [comment]
Looks as though something times out: 
``` 
Start 2: netcdfio_test
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

Do you think this is related to the PR, or is it something related to netcdf?

**Comment by setmar** [comment]
[Merge/implement variable dt for KP07 scheme](https://trello.com/c/q1L8WRhP/63-merge-implement-variable-dt-for-kp07-scheme)
